Basic/DyanmicPrefabs/Assets/Scripts/01_Server authoritative dynamic prefab loading across peers/DynamicPrefabManager.cs

```diff
@@ -0,0 +1,476 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Unity.Collections;
+using Unity.Netcode;
+using Unity.Netcode.Transports.UTP;
+using UnityEngine;
+using UnityEngine.AddressableAssets;
+using UnityEngine.ResourceManagement.AsyncOperations;
+
+namespace Game
+{
+    public sealed class DynamicPrefabManager : NetworkBehaviour
+    {
+        const int k_MaxConnectPayload = 1024;
+        const int k_EmptyDynamicPrefabHash = -1;
+        [SerializeField] float m_SpawnTimeoutInSeconds;
+        [SerializeField] NetworkManager m_NetworkManager;
+        
+        int m_CountOfClientsThatLoadedThePrefab = 0;
+        float m_SpawnTimeoutTimer = 0;
+        
+        int m_HashOfDynamicPrefabGUIDs = k_EmptyDynamicPrefabHash;
+        Dictionary<AddressableGUID, AsyncOperationHandle<GameObject>> m_LoadedDynamicPrefabResourceHandles = new Dictionary<AddressableGUID, AsyncOperationHandle<GameObject>>();
+        List<AddressableGUID> m_DynamicPrefabGUIDs = new List<AddressableGUID>(); //cached list to avoid GC
+
+        public event Action<ConnectStatus, FastBufferReader> OnConnectionStatusReceived;
+
+        string m_ConnectAddress;
+        ushort m_Port;
+        
+        //A storage where we keep the association between the dynamic prefab (hash of it's GUID) and the clients that have it loaded
+        Dictionary<int, HashSet<ulong>> m_PrefabHashToClientIds = new Dictionary<int, HashSet<ulong>>();
+        
+        //A storage where we keep association between prefab (hash of it's GUID) and the spawned network objects that use it
+        Dictionary<int, HashSet<NetworkObject>> m_PrefabHashToNetworkObjectId = new Dictionary<int, HashSet<NetworkObject>>();
+        
+        public bool HasClientLoadedPrefab(ulong clientId, int prefabHash) => m_PrefabHashToClientIds.TryGetValue(prefabHash, out var clientIds) && clientIds.Contains(clientId);
+
+        void RecordThatClientHasLoadedAPrefab(int assetGuidHash, ulong clientId)
+        {
+            if (m_PrefabHashToClientIds.TryGetValue(assetGuidHash, out var clientIds))
+            {
+                clientIds.Add(clientId);
+            }
+            else
+            {
+                m_PrefabHashToClientIds.Add(assetGuidHash, new HashSet<ulong>() {clientId});
+            }
+        }
+        
+        public void StartClient(string connectAddress, ushort port)
+        {
+            Debug.Log(nameof(StartClient));
+            m_ConnectAddress = connectAddress;
+            m_Port = port;
+            
+            m_NetworkManager.NetworkConfig.ForceSamePrefabs = false;
+            
+            var payload = JsonUtility.ToJson(new ConnectionPayload()
```

Comment:
@pdeschain This won't be useful for all techniques we're teaching about correct? Since this class seems to mix multiple cases together, it'd be great if you could identify which code is for which techniques to help @fernando-cortez  split things out later

Reply:
Well, I think that all of them are actually related - I wouldn't split them based on purely the fact that they show slightly different loading strategies - these are for different things I would think, but all o them are still part of the dynamic prefab management.

Reply:
And the connection payload is needed for any scenario that wishes to handle late joiners, which is I would think all of them.